src/tracing/agent.h

```diff
@@ -31,24 +32,51 @@ class TracingController : public v8::platform::tracing::TracingController {
   }
 };
 
+class Agent;
 
-class Agent {
+// Allows agent to be destroyed early. Methods are noop after the agent was
+// destroyed
+class AgentHandle {
  public:
-  // Resetting the pointer disconnects client
-  using ClientHandle = std::unique_ptr<std::pair<Agent*, int>,
-                                       void (*)(std::pair<Agent*, int>*)>;
+  explicit AgentHandle(Agent*);
+  void Reset();
+  void AppendTraceEvent(TraceObject* trace_event);
+  void Flush(bool blocking);
+  void Disconnect(int client);
 
-  static ClientHandle EmptyClientHandle() {
-    return ClientHandle(nullptr, DisconnectClient);
+ private:
+  Mutex mutex_;
+  Agent* agent_;
+};
+
+class ClientHandle {
+ public:
+  ClientHandle(std::shared_ptr<AgentHandle> agent, int id)
+               : agent_(agent), id_(id) {}
+  ClientHandle(const ClientHandle&) = delete;
+  ClientHandle& operator=(const ClientHandle&) = delete;
+  ~ClientHandle();
+  int id() const {
+    return id_;
   }
```

Comment:
The style guide allows doing this on one line.

Reply:
Done.